File: Java/Eta/Core/src/main/java/com/refinitiv/eta/codec/UInt.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2019-2022 Refinitiv. All rights reserved.         --
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.eta.codec;

import java.math.BigInteger;

/**
 * ETA Unsigned Integer type. Can currently represent an unsigned value with precision of up to 64 bits.
 */
public interface UInt
{
    /**
     * Clears {@link UInt}.
     */
    public void clear();
    
    /**
     * This method will perform a deep copy into the passed in parameter's 
     *          members from the Object calling this method.
     * 
     * @param destUInt the value getting populated with the values of the calling Object
     *
     * @return {@link CodecReturnCodes#SUCCESS} on success,
     *         {@link CodecReturnCodes#INVALID_ARGUMENT} if the destUInt is null. 
    */
    public int copy(UInt destUInt);
    
    /**
     * Is UInt blank.
     *
     * @return true, if is blank
     */
    public boolean isBlank();

    /**
     * Returns value as long. The value is returned as signed long. If the value
     * is negative, the user has to convert it to an unsigned value.
     * Alternatively, the user could call {@link #toBigInteger}.
     * 
     * @return the value as long
     */
    public long toLong();

    /**
     * Returns value as unsigned BigInteger.
     * This method may introduce garbage collection, as it creates BigInteger.
     * To avoid it, user may instead call the toLong method and implement the conversion.
     * 
     * @return the value as BigInteger
     */
    public BigInteger toBigInteger();

    /**
     * Set value to int.
     * 
     * @param value the value to set
     */
    public void value(int value);

    /**
     * Set value to long.
     * 
     * @param value the value to set
     */
    public void value(long value);

    /**
     * Set value to BigInteger. Must be in the range of 0 - (2^64-1).
     * 
     * @param value the value to set
     * 
     * @return {@link CodecReturnCodes#SUCCESS} on success,
     *         {@link CodecReturnCodes#INVALID_ARGUMENT} if value is invalid. 
     */
    public int value(BigInteger value);

    /**
     * Used to encode into a buffer.
     * 
     * @param iter {@link EncodeIterator} with buffer to encode into. Iterator
     *            should also have appropriate version information set.
     * 
     * @return {@link CodecReturnCodes}
     * 
     * @see EncodeIterator
     */
    public int encode(EncodeIterator iter);

    /**
     * Decode {@link UInt}.
     * 
     * @param iter {@link DecodeIterator} with buffer to decode from and appropriate
     *            version information set.
     * 
     * @return {@link CodecReturnCodes#SUCCESS} if success, {@link CodecReturnCodes#INCOMPLETE_DATA} if
     *         failure, {@link CodecReturnCodes#BLANK_DATA} if data is blank value
     * 
     * @see DecodeIterator
     */
    public int decode(DecodeIterator iter);

    /**
     * Convert {@link UInt} to a numeric string.
     * 
     * @return the string representation of this {@link UInt}
     */
    public String toString();

    /**
     * Set value to uint.
     * 
     * @param value the value to set
     * 
     * @return {@link CodecReturnCodes#SUCCESS} on success,
     *         {@link CodecReturnCodes#INVALID_ARGUMENT} if value is invalid. 
     */
    public int value(String value);
    
    /**
     * Checks equality of two UInt data.
     * 
     * @param thatUInt the other UInt to compare to this one
     * 
     * @return true if UInts are equal, false otherwise
     */
    public boolean equals(UInt thatUInt);

}

File: CSharp/Eta/Src/Codec/UInt.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2022 Refinitiv. All rights reserved.            --
 *|-----------------------------------------------------------------------------
 */

using System;

namespace Refinitiv.Eta.Codec
{
    /// <summary>
    /// ETA Unsigned Integer type. Can currently represent an unsigned value with precision of up to 64 bits.
    /// </summary>

    sealed public class UInt : IEquatable<UInt>
	{
        internal long _longValue;
		internal byte[] _bytes = new byte[9];

		// for value(String) method
		private string trimmedVal;

        /// <summary>
		/// Creates <see cref="UInt"/>.
		/// </summary>
		/// <returns> UInt object
		/// </returns>
		/// <seealso cref="UInt"/>
		public UInt()
		{
		}

        /// <summary>
		/// Clears <seealso cref="UInt"/>.
		/// </summary>
		public void Clear()
		{
			_longValue = 0;
		}

        /// <summary>
		/// This method will perform a deep copy into the passed in parameter's 
		///          members from the Object calling this method.
		/// </summary>
		/// <param name="destUInt"> the value getting populated with the values of the calling Object
		/// </param>
		/// <returns> <seealso cref="CodecReturnCode.SUCCESS"/> on success,
		///         <seealso cref="CodecReturnCode.INVALID_ARGUMENT"/> if the destUInt is null.  </returns>
		public CodecReturnCode Copy(UInt destUInt)
		{
			if (null == destUInt)
			{
				return CodecReturnCode.INVALID_ARGUMENT;
			}

			destUInt._longValue = _longValue;
			destUInt.IsBlank = IsBlank;

			return CodecReturnCode.SUCCESS;
		}

        /// <summary>
        /// Is UInt blank.
        /// </summary>
        public bool IsBlank { get; internal set; }

        /// <summary>
        /// Returns value as long. The value is returned as signed long. If the value
        /// is negative, the user has to convert it to an unsigned value.
        /// </summary>
        /// <returns> the value as long </returns>
        public long ToLong()
		{
			return _longValue;
		}

        /// <summary>
		/// Set value to int.
		/// </summary>
		/// <param name="value"> the value to set </param>
		public void Value(int value)
		{
			_longValue = value;
			IsBlank = false;
		}


        /// <summary>
        /// Set value to long.
        /// </summary>
        /// <param name="value"> the value to set </param>
        public void Value(long value)
		{
			_longValue = value;
			IsBlank = false;
		}

        /// <summary>
		/// Used to encode into a buffer.
		/// </summary>
		/// <param name="iter"> <see cref="EncodeIterator"/> with buffer to encode into. Iterator
		///            should also have appropriate version information set.
		/// </param>
		/// <returns> <see cref="CodecReturnCode"/>
		/// </returns>
		/// <seealso cref="EncodeIterator"/>
		public CodecReturnCode Encode(EncodeIterator iter)
		{
			if (!IsBlank)
			{
				return Encoders.PrimitiveEncoder.EncodeUInt((EncodeIterator)iter, this);
			}
			else
			{
				return CodecReturnCode.INVALID_ARGUMENT;
			}
		}

        /// <summary>
		/// Decode <seealso cref="UInt"/>.
		/// </summary>
		/// <param name="iter"> <see cref="DecodeIterator"/> with buffer to decode from and appropriate
		///            version information set.
		/// </param>
		/// <returns> <see cref="CodecReturnCode.SUCCESS"/> if success, <see cref="CodecReturnCode.INCOMPLETE_DATA"/> if
		///         failure, <see cref="CodecReturnCode.BLANK_DATA"/> if data is blank value
		/// </returns>
		/// <seealso cref="DecodeIterator"/>
		public CodecReturnCode Decode(DecodeIterator iter)
		{
			return Decoders.DecodeUInt(iter, this);
		}

        /// <summary>
        /// Convert <seealso cref="UInt"/> to a numeric string.
        /// </summary>
        /// <returns> the string representation of this <seealso cref="UInt"/> </returns>
        public override string ToString()
		{
			return Convert.ToString(_longValue);
		}

        /// <summary>
        /// Set value to uint.
        /// </summary>
        /// <param name="value"> the value to set
        /// </param>
        /// <returns> <seealso cref="CodecReturnCode.SUCCESS"/> on success,
        ///         <seealso cref="CodecReturnCode.INVALID_ARGUMENT"/> if value is invalid.  </returns>
        public CodecReturnCode Value(string value)
		{
			if (string.ReferenceEquals(value, null))
			{
				return CodecReturnCode.INVALID_ARGUMENT;
			}

			trimmedVal = value.Trim();
			if (trimmedVal.Length == 0)
			{
				// blank
				Blank();
				return CodecReturnCode.SUCCESS;
			}

			try
			{
				Value(long.Parse(value));
				return CodecReturnCode.SUCCESS;
			}
			catch (System.FormatException)
			{
				return CodecReturnCode.INVALID_ARGUMENT;
			}
		}

        /// <summary>
		/// Checks equality of two UInt data.
		/// </summary>
		/// <param name="thatUInt"> the other UInt to compare to this one
		/// </param>
		/// <returns> true if UInts are equal, false otherwise </returns>
		public bool Equals(UInt thatUInt)
		{
			return ((thatUInt != null) && (_longValue == thatUInt.ToLong()) && (IsBlank == thatUInt.IsBlank));
		}

        /// <summary>Determines whether the specified <c>Object</c> is equal to the current <c>Object</c>.</summary>
        /// <param name="other">The <c>Object</c> to compare with the this <c>Object</c>.</param>
        /// <returns><c>true</c> if the specified <c>Object</c> is equal to the current <c>Object</c>;
        /// otherwise, <c>false</c>.</returns>
        public override bool Equals(object other)
        {
            if (other == null)
                return false;

            if (other is Enum)
                return Equals((Enum)other);

            return false;
        }

        /// <summary>Serves as a hash function for a particular type.</summary>
        /// <returns>A hash code for the current <c>Object</c>.</returns>
        public override int GetHashCode()
        {
            return ToString().GetHashCode();
        }

        internal void Blank()
        {
            Clear();
            IsBlank = true;
        }
    }

}